pyproject.toml

```diff
@@ -1,7 +1,70 @@
 [build-system]
-requires = ["setuptools>=60", "setuptools_scm[toml]>=7.0"]
+requires = [
+    "setuptools==69.0.0",
+    "setuptools_scm==8.0.4",
```

Comment:
Why the pins here instead of minimum versions like before? Surely we support more than just these exact versions.

(The [setuptools pyproject.toml docs](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html) say that at least setuptools 61.0.0 is required.

Reply:
Reproducibility. For example, the `dynamic` argument in pyproject.toml wasn't required on setuptools 68.0, but it *is* required on 69.0. If you use an open ended pin, a project configuration that works on the day of release can break in the future.

With a hard pin, we know the configuration works with this specific set of versions, and that won't change. Those versions might not be compatible with future versions of pip, Python etc - but then, we can't make any guarantees about what might happen in future regardless.

Reply:
To clarify further: build-requires specify the environment that will be created specifically for the purpose of wheel creation. It's guaranteed to be an isolated environment (unless you specifically disable isolation, in which case you're on your own), so we don't need to maintain maximum compatibility with the end-user's environment. You're specifying the exact environment that you want to be present to build the wheel; an open ended pin undermines that, because we can't guarantee that the build environment will match a configuration that we know works.